packages/flutter/lib/src/painting/box_border.dart

```diff
@@ -603,50 +633,56 @@ class Border extends BoxBorder {
       }
     }
 
-    // Allow painting non-uniform borders if the color and style are uniform.
-    if (_colorIsUniform && _styleIsUniform) {
-      switch (top.style) {
-        case BorderStyle.none:
-          return;
-        case BorderStyle.solid:
-          BoxBorder._paintNonUniformBorder(canvas, rect,
-              shape: shape,
-              borderRadius: borderRadius,
-              textDirection: textDirection,
-              left: left,
-              top: top,
-              right: right,
-              bottom: bottom);
-          return;
-      }
+    if (_styleIsUniform && top.style == BorderStyle.none) {
+      return;
     }
 
-    assert(() {
-      if (borderRadius != null) {
+    // Allow painting non-uniform borders if the visible colors are uniform.
+    final Set<Color> visibleColors = _distinctVisibleColors;
+    final bool hasHairlineBorder = _hasHairlineBorder;
+    // Paint a non uniform border if a single color is visible
+    // and (borderRadius is present) or (border is visible and width != 0.0).
+    if (visibleColors.length == 1 && !hasHairlineBorder) {
+      BoxBorder.paintNonUniformBorder(canvas, rect,
+          shape: shape,
+          borderRadius: borderRadius,
+          textDirection: textDirection,
+          top: top.style == BorderStyle.none ? null : top,
+          right: right.style == BorderStyle.none ? null : right,
+          bottom: bottom.style == BorderStyle.none ? null : bottom,
+          left: left.style == BorderStyle.none ? null : left,
+          color: visibleColors.first);
+      return;
+    }
+
+     assert(() {
+      if (hasHairlineBorder) {
+        assert(borderRadius == null || borderRadius == BorderRadius.zero,
+            'A hairline border like `BorderSide(width: 0.0, style: BorderStyle.solid)` can only be drawn when BorderRadius is zero or null.');
+      }
```

Comment:
why can't we render hairline rounded rectangles?

Reply:
Because the way we paint hairline is setting the paint style to stroke, which breaks everything; it will never work with `drrect`.

But... there is an alternative.
We could use 1 / density to get 1px, and therefore support in drrect. But.... 1 / density requires the context, and Border doesn't have context, so it is not possible.

It is easier to just deprecate the hairline behavior and invent a `PixelBorder(leftWidth, ..., context)`.